### PR TITLE
CLN: Deprecate dayofweek/hello day_of_week (#9606)

### LINKED ISSUE #9606: API: timeseries accessors naming convention
we have a slightly odd naming convention for several index (and .dt) accessors

`dayofweek`
`dayofyear`

(and adding `daysinmonth` in #9605)

so these should prob be `day_of_week` 

we should be able to maintain compat on these and just add the new ones.

### PR DESCRIPTION
**Pull request by abhishekmangla**
Rename dayofweek properties in Period, Datetime, and PeriodArray classes.
Add day_of_week property to respective tests for each class.

- [x] closes #9606
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/_libs/tslibs/timestamps.pyx [hunk 1]
```diff
@@ -538,12 +538,19 @@ cdef class _Timestamp(ABCTimestamp):
...
 
     @property
-    def dayofweek(self) -> int:
+    def day_of_week(self) -> int:
         """
         Return day of the week.
         """
         return self.weekday()
 
+    @property
```
Comment (arw2019) [inline pandas/_libs/tslibs/timestamps.pyx:547]: You don't need to define a second property, set `dayofweek` to be an alias of `day_of_week`

File: pandas/_libs/tslibs/period.pyx [hunk 2]
```diff
@@ -1900,38 +1900,43 @@ cdef class _Period(PeriodMixin):
...
+        >>> per.day_of_week
         2
-        >>> per.end_time.dayofweek
+        >>> per.end_time.day_of_week
         2
         """
         base = self._dtype._dtype_code
         return pweekday(self.ordinal, base)
 
+    @property
```
Comment (arw2019) [inline pandas/_libs/tslibs/period.pyx:1935]: I don't think you need this? Just set dayofweek` to be an alias of `day_of_week`

File: pandas/tests/generic/test_finalize.py [hunk 3]
```diff
@@ -678,6 +678,8 @@ def test_datetime_method(method):
         "microsecond",
         "nanosecond",
         "dayofweek",
+        # Add better named field for GH-9606
```
Comment (arw2019) [inline pandas/tests/generic/test_finalize.py:681]: nit-pick but we usually don't leave comments like this: git/blame keeps track of when changes occurred

File: pandas/tests/indexes/datetimes/test_scalar_compat.py [hunk 4]
```diff
@@ -38,6 +38,8 @@ def test_dti_date_out_of_range(self, data):
         "field",
         [
             "dayofweek",
+            # Add better named field for GH-9606
```
Comment (arw2019) [inline pandas/tests/indexes/datetimes/test_scalar_compat.py:41]: same here

File: pandas/tests/indexes/period/test_period.py [hunk 5]
```diff
@@ -249,6 +249,8 @@ def _check_all_fields(self, periodindex):
             "weekofyear",
             "week",
             "dayofweek",
+            # Add better named field for GH-9606
```
Comment (arw2019) [inline pandas/tests/indexes/period/test_period.py:252]: same here

**Review by arw2019 — CHANGES_REQUESTED** [review]
Thanks @abhishekmangla for the PR!

Looks good, some comments.

You'll also need to add a whatsnew (v1.2) under enhancements

Let us know if you have any questions!

**Comment by abhishekmangla** [comment]
@arw2019 Thank you for the review. I addressed your comments and added an entry to whatsnew file.

File: doc/source/whatsnew/v1.2.0.rst [hunk 6]
```diff
@@ -13,6 +13,21 @@ including other versions of pandas.
 Enhancements
 ~~~~~~~~~~~~
 
+.. _whatsnew_120.improve_timeseries_accessor_naming:
```
Comment (arw2019) [inline doc/source/whatsnew/v1.2.0.rst:16]: sorry to be a pain but could you do something short here like the whatsnew in this PR:
 https://github.com/pandas-dev/pandas/pull/9605/files

File: pandas/_libs/tslibs/period.pyx [hunk 7]
```diff
@@ -1475,6 +1475,9 @@ cdef class _Period(PeriodMixin):
         PeriodDtypeBase _dtype
         BaseOffset freq
 
+    # Alias dayofweek function to day_of_week (GH-9606)
```
Comment (arw2019) [inline pandas/_libs/tslibs/period.pyx:1478]: let's drop this comment

File: pandas/_libs/tslibs/timestamps.pyx [hunk 8]
```diff
@@ -230,6 +230,8 @@ cdef class _Timestamp(ABCTimestamp):
 
     # higher than np.ndarray and np.matrix
     __array_priority__ = 100
+    # Alias dayofweek function to day_of_week (GH-9606)
```
Comment (arw2019) [inline pandas/_libs/tslibs/timestamps.pyx:233]: Same here, let's drop this comment

**Review by arw2019 — CHANGES_REQUESTED** [review]
Thanks for following up on this!

Some more (minor) comments, otherwise good to go IMO

**Comment by abhishekmangla** [comment]
@arw2019 Resolved minor issues. Do you know how to locally test .rst files, in the CI stage of the checks, I am failing things like
`/home/runner/work/pandas/pandas/pandas/core/indexes/period.py:docstring of pandas.PeriodIndex.rst:77: WARNING: autosummary: stub file not found 'pandas.PeriodIndex.day_of_year'. Check your autosummary_generate setting.
2391`
AND
`/home/runner/work/pandas/pandas/doc/source/whatsnew/v1.2.0.rst:31: WARNING: Explicit markup ends without a blank line; unexpected unindent.`

`./ci/code_checks.sh` shows no errors locally

**Comment by arw2019** [comment]
> @arw2019 Resolved minor issues. Do you know how to locally test .rst files? In the CI stage of the checks, I am failing things like
> `/home/runner/work/pandas/pandas/pandas/core/indexes/period.py:docstring of pandas.PeriodIndex.rst:77: WARNING: autosummary: stub file not found 'pandas.PeriodIndex.day_of_year'. Check your autosummary_generate setting. 2391`

I think you (maybe) need to add the new methods to the rst files manually - for example in doc/source/reference/series.rst both    daysinmonth and days_in_month are listed

> `/home/runner/work/pandas/pandas/doc/source/whatsnew/v1.2.0.rst:31: WARNING: Explicit markup ends without a blank line; unexpected unindent.`

This one is a linting issue

File: doc/source/whatsnew/v1.2.0.rst [hunk 9]
```diff
@@ -13,14 +13,23 @@ including other versions of pandas.
 Enhancements
 ~~~~~~~~~~~~
 
+
+
+.. _whatsnew_120.improve_timeseries_accessor_naming:
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:18]: you can just put these notes in Other Enhancements section

File: doc/source/whatsnew/v1.2.0.rst [hunk 10]
```diff
@@ -13,14 +13,23 @@ including other versions of pandas.
...
+
 .. _whatsnew_120.duplicate_labels:
 
 Optionally disallow duplicate labels
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-:class:`Series` and :class:`DataFrame` can now be created with ``allows_duplicate_labels=False`` flag to
-control whether the index or columns can contain duplicate labels (:issue:`28394`). This can be used to
-prevent accidental introduction of duplicate labels, which can affect downstream operations.
+:class:`Series` and :class:`DataFrame` can now be created with ``allows_duplicate_labels=False`` flag to control whether the index or columns can contain duplicate labels (:issue:`28394`). This can be used to prevent accidental introduction of duplicate labels, which can affect downstream operations.
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:32]: pls don't adjust other notes

**Comment by abhishekmangla** [comment]
@arw2019 The only unit test that is failing on Azure pipeline is `test_binary_arith_ops` which seems to be a known unreliable test (#37328). (The test suite passes on my macOS machine)

Is this good to merge now?

**Review by jreback — CHANGES_REQUESTED** [review]
this is fine. ping on green.

**Comment by jreback** [comment]
@pandas-dev/pandas-core should we do a DeprecationWarning / FutureWarning on the original attributes? (would be a followon)

**Comment by abhishekmangla** [comment]
@jreback All checks green except aforementioned `test_binary_arith_ops`

**Comment by jreback** [comment]
thanks @abhishekmangla very nice

**Pull request merged**